04-higher-order/homework-04/hw-04.hs

```diff
@@ -0,0 +1,177 @@
+and so on. A binary tree is balanced if the height of its left and right
+subtrees differ by no more than 1, and its left and right subtrees are
+also balanced.
+You should use the following data structure to represent binary
+trees. Note that each node stores an extra Integer representing the
+height at that node.
+data Tree a = Leaf
+| Node Integer (Tree a) a (Tree a)
+deriving (Show, Eq)
+For this exercise, write a function
+cis 194: homework 4 2
+foldTree :: [a] -> Tree a
+foldTree = ...
+which generates a balanced binary tree from a list of values using
+foldr.
+For example, one sample output might be the following, also visualized at right:
+D
+E
+A
+G
+H
+B
+F C
+I
+J
+foldTree "ABCDEFGHIJ" ==
+Node 3
+(Node 2
+(Node 0 Leaf ’F’ Leaf)
+’I’
+(Node 1 (Node 0 Leaf ’B’ Leaf) ’C’ Leaf))
+’J’
+(Node 2
+(Node 1 (Node 0 Leaf ’A’ Leaf) ’G’ Leaf)
+’H’
+(Node 1 (Node 0 Leaf ’D’ Leaf) ’E’ Leaf))
+Your solution might not place the nodes in the same exact order,
+but it should result in balanced trees, with each subtree having a
+correct computed height.
+-}
+--Exercise 2
+data Tree a = Leaf | Node Integer (Tree a) a (Tree a)
+    deriving (Show, Eq)
+
+foldTree :: [a] -> Tree a
+foldTree = foldr insert Leaf
+  where 
+    insert elem Leaf = Node 0 Leaf elem Leaf
+    insert elem (Node hgt left val right)
+        | getHgt left > getHgt right = Node hgt                                        left val (insert elem right)
+        | getHgt left < getHgt right = Node hgt                                        (insert elem left) val right
+        | otherwise                  = Node (1 + (getMaxHgt right $ insert elem left)) (insert elem left) val right
+          where
+            getHgt Leaf                    = -1
+            getHgt (Node hgt _ _ _)        = hgt
+            getMaxHgt firstTree secondTree = max (getHgt firstTree) (getHgt secondTree)
+
+
+{-
+Exercise 3: More folds!
+1. Implement a function
+xor :: [Bool] -> Bool
+which returns True if and only if there are an odd number of True
+values contained in the input list. It does not matter how many
+False values the input list contains. For example,
+xor [False, True, False] == True
+xor [False, True, False, False, True] == False
+Your solution must be implemented using a fold.
+2. Implement map as a fold. That is, complete the definition
+map’ :: (a -> b) -> [a] -> [b]
+map’ f = foldr ...
+in such a way that map’ behaves identically to the standard map
+function.
+cis 194: homework 4 3
+3. (Optional) Implement foldl using foldr. That is, complete the
+definition
+myFoldl :: (a -> b -> a) -> a -> [b] -> a
+myFoldl f base xs = foldr ...
+in such a way that myFoldl behaves identically to the standard
+foldl function.
+Hint: Study how the application of foldr and foldl work out:
+foldr f z [x1, x2, ..., xn] == x1 ‘f‘ (x2 ‘f‘ ... (xn ‘f‘ z)...)
+foldl f z [x1, x2, ..., xn] == (...((z ‘f‘ x1) ‘f‘ x2) ‘f‘...) ‘f‘ xn
+-}
+--Exercise 3
+xor :: [Bool] -> Bool
+xor = foldr (/=) False
```

Comment:
```suggestion
-- import Data.Foldable (foldl')
xor = foldl' (/=) False
```

If we run benchmarks:

```
benchmarking all true/foldr
time                 95.74 ms   (85.65 ms .. 101.8 ms)
                     0.989 R²   (0.970 R² .. 1.000 R²)
mean                 91.96 ms   (80.80 ms .. 100.4 ms)
std dev              15.69 ms   (9.977 ms .. 26.64 ms)
variance introduced by outliers: 54% (severely inflated)

benchmarking all true/foldl'
time                 7.265 ms   (7.158 ms .. 7.366 ms)
                     0.999 R²   (0.998 R² .. 0.999 R²)
mean                 7.183 ms   (7.126 ms .. 7.249 ms)
std dev              171.7 μs   (150.0 μs .. 204.9 μs)

benchmarking all false/foldr
time                 22.12 ms   (21.67 ms .. 22.57 ms)
                     0.999 R²   (0.997 R² .. 1.000 R²)
mean                 22.69 ms   (22.48 ms .. 22.92 ms)
std dev              497.0 μs   (378.5 μs .. 731.1 μs)

benchmarking all false/foldl'
time                 6.612 ms   (6.491 ms .. 6.771 ms)
                     0.997 R²   (0.995 R² .. 0.999 R²)
mean                 6.520 ms   (6.466 ms .. 6.602 ms)
std dev              195.0 μs   (143.0 μs .. 247.2 μs)
variance introduced by outliers: 13% (moderately inflated)

benchmarking alternating/foldr
time                 65.99 ms   (58.06 ms .. 71.84 ms)
                     0.968 R²   (0.913 R² .. 0.996 R²)
mean                 63.10 ms   (56.35 ms .. 68.85 ms)
std dev              11.45 ms   (8.236 ms .. 15.77 ms)
variance introduced by outliers: 62% (severely inflated)

benchmarking alternating/foldl'
time                 7.647 ms   (7.471 ms .. 7.835 ms)
                     0.995 R²   (0.991 R² .. 0.998 R²)
mean                 7.501 ms   (7.393 ms .. 7.632 ms)
std dev              336.0 μs   (261.2 μs .. 432.2 μs)
variance introduced by outliers: 22% (moderately inflated)
```

It appears that using `foldl'` is significantly faster than `foldr`. Why is this so?

[benchmark script](https://gist.github.com/kwannoel/567d7795c67907ce1bb9a4cbc086e8bc).

Reply:
Once you figure out, there are other areas in your code in which using `foldl'` will improve performance further.

Reply:
@kwannoel Thanks for explaining the differences in speed using the benchmark tests.